sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/GenericArrayData.scala

```diff
@@ -142,3 +164,415 @@ class GenericArrayData(val array: Array[Any]) extends ArrayData {
     result
   }
 }
+
+final class GenericIntArrayData(private val primitiveArray: Array[Int])
```

Comment:
Thank you for your suggestion. It would be good to make ```GenericArrayData``` abstract.
As you said, ```ArrayData``` will be mega-morphic in the class hierarchy, actual use should be monomorphic. To make each class ```final``` alleviate runtime overhead.